File: 211_add_and_search_word-data_structure_design.rb
=begin
Design a data structure that supports the following two operations:

void addWord(word)
bool search(word)
search(word) can search a literal word or a regular expression string containing only letters a-z or .. A . means it can represent any one letter.

Example:

addWord("bad")
addWord("dad")
addWord("mad")
search("pad") -> false
search("bad") -> true
search(".ad") -> true
search("b..") -> true
Note:
You may assume that all words are consist of lowercase letters a-z.
=end

class WordDictionary

=begin
    Initialize your data structure here.
=end
    def initialize()
        @root = TrieNode.new
    end

=begin
    Adds a word into the data structure.
    :type word: String
    :rtype: Void
=end
    def add_word(word)
        node = @root
        word.each_char do |char|
            letter_index = char.ord - 'a'.ord
            node.children[letter_index] ||= TrieNode.new
            node = node.children[letter_index]
        end
        node.is_word = true
    end


=begin
    Returns if the word is in the data structure. A word could contain the dot character '.' to represent any one letter.
    :type word: String
    :rtype: Boolean
=end
    def search(word)
        match(word, 0, @root)
    end
    
    private
    
    def match(word, i, node)
        if i == word.length
            return node.is_word
        end
        if word[i] != '.'
            letter_index = word[i].ord - 'a'.ord
            child_node = node.children[letter_index]
            if child_node
                match(word, i+1, child_node)
            else
                return false
            end
        else
            (0...node.children.length).each do |letter_index|
                child_node = node.children[letter_index]
                if child_node
                    if match(word, i+1, child_node)
                        return true
                    end
                end
            end
            
            return false
        end
    end
end

# Your WordDictionary object will be instantiated and called as such:
# obj = WordDictionary.new()
# obj.add_word(word)
# param_2 = obj.search(word)

class TrieNode
    attr_accessor :children, :is_word
    def initialize
        @children = Array.new(26)
        @is_word = false
    end
end

File: 211_add_and_search_word-data_structure_design.java
/*
Design a data structure that supports adding new words and finding if a string matches any previously added string.

Implement the WordDictionary class:

WordDictionary() Initializes the object.
void addWord(word) Adds word to the data structure, it can be matched later.
bool search(word) Returns true if there is any string in the data structure that matches word or false otherwise. word may contain dots '.' where dots can be matched with any letter.
 

Example:

Input
["WordDictionary","addWord","addWord","addWord","search","search","search","search"]
[[],["bad"],["dad"],["mad"],["pad"],["bad"],[".ad"],["b.."]]
Output
[null,null,null,null,false,true,true,true]

Explanation
WordDictionary wordDictionary = new WordDictionary();
wordDictionary.addWord("bad");
wordDictionary.addWord("dad");
wordDictionary.addWord("mad");
wordDictionary.search("pad"); // return False
wordDictionary.search("bad"); // return True
wordDictionary.search(".ad"); // return True
wordDictionary.search("b.."); // return True
 

Constraints:

1 <= word.length <= 500
word in addWord consists lower-case English letters.
word in search consist of  '.' or lower-case English letters.
At most 50000 calls will be made to addWord and search.
*/

class WordDictionary {
    Trie trie;

    public WordDictionary() {
        this.trie = new Trie();
    }
    
    public void addWord(String word) {
        trie.add(word);
    }
    
    public boolean search(String word) {
        return trie.search(word);
    }
}

class Trie {
    Trie[] children;
    boolean isWord;
    Trie(){ this.children = new Trie[26];}
    
    
    public void add(String word) {
        Trie curr = this;
        for(char c : word.toCharArray()){
            if(curr.children[c - 'a'] == null) curr.children[c - 'a'] = new Trie();
            curr = curr.children[c - 'a'];
        }
        curr.isWord = true;
    }
    
    public boolean search(String word) {
        return search(this, word, 0);
    }
    
    private boolean search(Trie curr, String word, int idx) {
        if (idx == word.length()) {
            return curr.isWord;
        }
        char c = word.charAt(idx);
        
        if (c != '.') {
            return curr.children[c - 'a'] != null && search(curr.children[c-'a'], word, idx+1);
        }
        
        for (int i = 0; i < 26; i++) {
            if (curr.children[i] != null) {
                if (search(curr.children[i], word, idx + 1)) {
                    return true;
                }
            }
        }
        return false;
    }
}

/**
 * Your WordDictionary object will be instantiated and called as such:
 * WordDictionary obj = new WordDictionary();
 * obj.addWord(word);
 * boolean param_2 = obj.search(word);
 */

